app/main/controllers/clusters.py

```diff
@@ -0,0 +1,225 @@
+    source = Source.query.get(source_id)
+    if source is None:
+        abort(404)
+
+    if source.project.user_id != user_id:
+        raise AuthError({
+            'code': 'invalid_user',
+            'description': 'This item does not belong to the requesting user.'
+        }, 403)
+
+    return source
+
+
+def get_authorized_item(user_id, item_id):
+    item = Item.query.get(item_id)
+
+    if item is None:
+        abort(404)
+    if item.project.user_id != user_id:
+        raise AuthError({
+            'code': 'invalid_user',
+            'description': 'This item does not belong to the requesting user.'
+        }, 403)
+
+    return item
+
+
+def set_cluster_routes(app):
+    '''
+    Displays clusters and items of a cluster
+    '''
+    @app.route('/clusters/<int:cluster_id>')
+    @requires_auth('read:clusters')
+    def get_cluster_info(user_id, cluster_id):
+        cluster = Cluster.query.filter(Cluster.id == cluster_id).first()
+        return jsonify({
+            'success': True,
+            'cluster': cluster.format()
+        })
+
+    """
+    Deletes a cluster.
+    """
+    @app.route('/clusters/<int:cluster_id>', methods=['DELETE'])
+    @requires_auth('delete:clusters')
+    def delete_cluster(user_id, cluster_id):
+        cluster = get_authorized_cluster(user_id, cluster_id)
+        # Set source parent cluster equal to the parent cluster
+        # of the cluster being deleted(could be none)
+        cluster_parent = cluster.parent_cluster
+        for item in cluster.child_items:
+            if cluster_parent is None:
+                # items now just belong to the project -
+                cluster.project.items.append(item)
+                item.cluster_id = None
+            else:
+                cluster_parent.child_items.append(item)
+                # I may not need to explicitly do this
+                item.cluster_id = cluster_parent.id
+            item.update()
+
+        if cluster_parent is None:
+            cluster.project.update()
+        else:
+            cluster_parent.update()
+        cluster.delete()
+
+        return jsonify({
+            'success': True,
+            'deleted': cluster_id
+        })
+
+    '''
+    Update cluster name
+    '''
+    @app.route('/clusters/<int:cluster_id>', methods=['PATCH'])
+    @requires_auth('update:clusters')
+    def update_cluster_name(user_id, cluster_id):
+        cluster = get_authorized_cluster(user_id, cluster_id)
+        body = request.get_json()
+        if body is None:
+            abort(400)
+
+        new_name = body.get('name', None)
+        if new_name is None:
+            abort(400)
+
+        cluster.name = new_name
+        cluster.update()
+
+        return jsonify({
+            'success': True,
+            'cluster': cluster.format()
+        })
+
+    '''
+    Creates a new cluster from scratch (two sources come near each other)
+    '''
+    @app.route('/clusters/create_new', methods=['POST'])
```

Comment:
Should just be a POST to `/clusters`, no need to add the `create_new`
```suggestion
    @app.route('/clusters', methods=['POST'])
```